ksql-engine/src/main/java/io/confluent/ksql/KsqlContext.java

```diff
@@ -85,13 +85,15 @@ public FunctionRegistry getFunctionRegistry() {
   }
 
   public List<QueryMetadata> sql(final String sql, final Map<String, Object> overriddenProperties) {
-    final List<PreparedStatement<?>> statements = ksqlEngine.parseStatements(sql);
+    final List<ParsedStatement> statements = ksqlEngine.parse(sql);
 
     final KsqlExecutionContext sandbox = ksqlEngine.createSandbox();
-    statements.forEach(stmt -> sandbox.execute(stmt, ksqlConfig, overriddenProperties));
+
+    statements.forEach(stmt ->
+        sandbox.execute(sandbox.prepare(stmt), ksqlConfig, overriddenProperties));
 
     final List<QueryMetadata> queries = statements.stream()
-        .map(stmt -> ksqlEngine.execute(stmt, ksqlConfig, overriddenProperties))
+        .map(stmt -> ksqlEngine.execute(ksqlEngine.prepare(stmt), ksqlConfig, overriddenProperties))
```

Comment:
I remember someone telling me recently that streams are lazily evaluated ;)

In this case the order of evaluation is important.  Consider two statements:

```sql
CS FOO;
CS BAR AS SELECT * FROM FOO;
```

It is important that we only `prepare` the second statement once the first has been executed. But this is not guaranteed with your change. 

Hence my explicit code.

Reply:
Oh man... I think we should probably avoid statements with side-effects in streams in general. Streams technically don't even guarantee ordering at all (sequential streams, the default, do but not by API contract) - it is possible that even with your code the second statements could be re-ordered.

If you look at the javadoc for Streams:

> A stream implementation is permitted significant latitude in optimizing the computation of the result. For example, a stream implementation is free to elide operations (or entire stages) from a stream pipeline -- and therefore elide invocation of behavioral parameters -- if it can prove that it would not affect the result of the computation. This means that side-effects of behavioral parameters may not always be executed and should not be relied upon, unless otherwise specified (such as by the terminal operations forEach and forEachOrdered). (For a specific example of such an optimization, see the API note documented on the count operation. For more detail, see the side-effects section of the stream package documentation.)
> ...
> Most stream operations accept parameters that describe user-specified behavior, such as the lambda expression w -> w.getWeight() passed to mapToInt in the example above. To preserve correct behavior, these behavioral parameters:
> - must be non-interfering (they do not modify the stream source); and
> - **in most cases must be stateless (their result should not depend on any state that might change during execution of the stream pipeline).**

My general opinion is that streams are awesome, but easy to abuse.

Reply:
Or better yet, it's explicitly mentioned here: https://docs.oracle.com/javase/8/docs/api/java/util/stream/package-summary.html 

> Side-effects in behavioral parameters to stream operations are, in general, discouraged, as they can often lead to unwitting violations of the statelessness requirement, as well as other thread-safety hazards.

Reply:
Luckily, we have tests to prove the code does what's intended. So we can rest easy.

Reply:
As a person who originally raised this issue I see there is a danger in confusing "what pure streams do" and the practicalities of using KSQL in the real world. The issue I started with is that:

- KSQL only permits A-B joining and not complex A-B-C-(n) joins. This means that quite often we need to define multiple KSQL streams or tables to get all of the data to be joined joined!
- Rekey requirements force us to create new streams from existing topics and streams

Thus we often end up with multiple create statements that implement some form of transformation pipeline. These steps have to be sequence; in CLI interactive mode a developer could just issue each create in order and all would work. In SQL scripted modes this is not possible as the metadata is incomplete and the whole scripted file fails.
If the parsing solution cannot be made to work for philosophical reasons the only viable alternative is to run multiple KSQL servers each handling a component of the total data pipeline - which is probably the wrong answer

Reply:
@DW-Pete - thanks for chiming in! My comment wasn't regarding the validity of the change, I agree that this change should be made - the philosophical debate is just whether we use java8 streams to code this or leave it to old-school iterators/for loops.

---

@big-andy-coates - I want to split concerns in two: ordering and side-effects. I totally agree with you that if you created the stream then you can probably assume that it is sequential and don't have to worry about possible unordered execution (otherwise lots of our assumptions would break down). On the other hand, I do think side-effects in streams should be avoided.

Avoiding intermediate side-effects is a pretty well established guideline in the functional programming world, and I think we should follow this guideline. Most of our usage of Java 8 Streams API has one clear benefit: compact code and readability (as with this example, we don't usually take advantage of the laziness or the concurrency benefits). Introducing side-effects severely detracts from this singular benefit: we now need to reason about laziness, partial streams execution, internal optimizations, synchrony etc... - all of these things that if we avoid side-effects I can just reason away by looking at the code. This is not to mention the debugging overhead and the ability to examine intermediary state, which becomes more important in critical state-changing code such as this, or all the other secondary issues with streams (performance, object allocations, etc...). 

And is the alternative really so bad? Using a for each loop in these situations is clear, concise and only adds one or two lines of overhead. And if we stuck to this pattern, it would make reviewing and maintaining code that uses streams way easier for every non-side-effect inducing code!

There are few things that I feel this strongly about, but this is something that I've thought a lot about - I think there are so many reasons to avoid it and, unless I'm missing something (which I'm happy to be enlightened about), I don't see a convincing reason to use streams in these cases.

Reply:
Hey @DW-Pete, 

- I hear you on the multi-way joins. We're aware of that as a high priority feature.
- Rekey topics are an artefact of using KStreams under the hood. There is some discussion of doing shuffles across the network, rather than via Kafka, though early days and nothing confirmed.  I'm interested to hear if you see the rekey topics as a big issue for you?

Reply:
In this context rekey is only potential problem if the source is not recognised because it is not yet "registered" somewhere in the metadata or parse process.

More generally, rekey creates stuff and stuff takes space .... I'm old and my programming career started in C and 6502 Assembler. I get cranky with bytes

Reply:
Hey @agavra,

This is a fun discussion :D. The zealot versus the pragmatist :p.

I would certainly concede that you can write streams code with side effects that doesn't do what you expect if you don't know what you're doing, due to their lazy nature.  But the same can be true of many things, e.g. threading, or lambas, yet we don't avoid using lambas for things with side effects.

I can see an argument for preferring the humble for loop so that people who don't know what they're doing with streams don't refactor this loop to do the wrong thing.  Though I personally find the proposition of dumbing down your code to the lowest common denominator a bit depressing and can actually result in more bugs, due to the code being more verbose and less expressive. Plus, this is one of the reasons we write tests: so that they break when someone breaks the code.

Also, how far do you take this no-side effect functional paradigm? I was looking at the code in question and rewrote it as:

```java
final List<QueryMetadata> queries = new ArrayList<>();
    
for (final ParsedStatement statement : statements) {
   final PreparedStatement<?> prepared = ksqlEngine.prepare(statement);
   final ExecuteResult result = ksqlEngine.execute(prepared, ksqlConfig, overriddenProperties);
   result.getQuery().ifPresent(queries::add);
}
```

But wait... `ifPresent` is a functional style operation... and in this code it has size effects!

So I still don't necessarily agree with the 'thou must never have side effects' matra. I prefer the more pragmatic: 'be careful / aware of side effects', just like you should be in all the programming we do.

All that being said - I think you're right in this case. (I think I was just arguing the toss for the sake of it ;) ). As the importance of the ordering in this code is key, and not immediately obvious, I'll switch it.

BTW, it's great to have these discussions, so continue to do what you're doing.  I appreciate a thorough code review.

Reply:
Actually, upon revisiting the code my preferred solution would just be to move the `prepare` and `execute` methods into a function. Making the code:

```java
statements.stream()
        .map(stmt -> execute(ksqlEngine, stmt, ksqlConfig, overriddenProperties))
        .map(ExecuteResult::getQuery)
        .filter(Optional::isPresent)
        .map(Optional::get)
        .collect(Collectors.toList());

private static ExecuteResult execute(
      final KsqlExecutionContext executionContext,
      final ParsedStatement stmt,
      final KsqlConfig ksqlConfig,
      final Map<String, Object> overriddenProperties
  ) {
    final PreparedStatement<?> prepared = executionContext.prepare(stmt);
    return executionContext.execute(prepared, ksqlConfig, overriddenProperties);
  }
```

However, as I'm not as passionate about this as your good self, and to get things moving forward, I'll still change the code.

Reply:
A few more thoughts:

> This is a fun discussion :D. The zealot versus the pragmatist :p.

I'm enjoying it too - please let me know if I ever get too annoying or stubborn!

> But wait... `ifPresent` is a functional style operation... and in this code it has size effects!

While I would love for us to migrate KSQL code to OCaml and herald in an era of functional glory, the unfortunate (or fortunate?) reality is that we live in a stateful world: _"without side effects, the only thing that programs could do is make machines run really hot"_. That means that at some point, even the most functional programming code needs to have side-effects, the question is being thoughtful of _where_ you put that side effects. IMO, "terminal" operations (e.g. `forEach`, `collect`, `ifPresent`) are totally valid places for side-effects to occur (in fact they _must_ happen in terminal operations, otherwise you get compiler warnings!).

> (though I'll still be using functional style programming with side effects in the new code :p).

I'm sure you'll catch me doing the same! To make it slightly less hypocritical, I suggest drawing a line at "is this a place I would want to put a debug breakpoint and examine the intermediate state?" - if yes, then it's probably better keeping it imperative.

Reply:
> I'm enjoying it too - please let me know if I ever get too annoying or stubborn!

Never! (and likewise)

> "is this a place I would want to put a debug breakpoint and examine the intermediate state?"

I have no issues adding breakpoints to streams code... not sure what you're doing wrong!?!? :p